pkg/driverbuilder/builder/opensuse.go

```diff
@@ -0,0 +1,253 @@
+package builder
+
+import (
+	_ "embed"
+	"fmt"
+	"strings"
+
+	"github.com/falcosecurity/driverkit/pkg/kernelrelease"
+)
+
+//go:embed templates/opensuse.sh
+var opensuseTemplate string
+
+// TargetTypeOpenSUSE identifies the OpenSUSE target.
+const TargetTypeOpenSUSE Type = "opensuse"
+
+// We need:
+// kernel-default-devel-*-{arch}
+// kernel-devel-*-noarch
+const opensuseMinimumURLs = 2
+
+// base URLs to begin searches
+var baseURLs []string = []string{
+	// general releases, leap releases
+	"https://mirrors.edge.kernel.org/opensuse/distribution",
+	"http://download.opensuse.org/distribution",
+	"https://download.opensuse.org/repositories/Kernel:",
+	// some releases are stored at the top level specifically
+	"http://download.opensuse.org",
+}
+
+// all known releases - will need to expand as more are added
+var releases []string = []string{
+	// openSUSE leap
+	"43.2",
+	"15.0",
+	"15.1",
+	"15.2",
+	"15.3",
+	"15.4",
+	"15.5",
+	"15.6",
+	// other releases
+	"HEAD",
+	"stable",
+	"tumbleweed",
+}
+
+func init() {
+	BuilderByTarget[TargetTypeOpenSUSE] = &opensuse{}
+}
+
+// opensuse is a driverkit target.
+type opensuse struct {
+}
+
+type opensuseTemplateData struct {
+	commonTemplateData
+	KernelDownloadURLs []string
+}
+
+func (o *opensuse) MinimumURLs() int {
+	return opensuseMinimumURLs
+}
+
+func (o *opensuse) Name() string {
+	return TargetTypeOpenSUSE.String()
+}
+
+func (o *opensuse) TemplateScript() string {
+	return opensuseTemplate
+}
+
+func (o *opensuse) URLs(_ Config, kr kernelrelease.KernelRelease) ([]string, error) {
+
+	// SUSE requires 2 urls: a kernel-default-devel*{arch}.rpm and a kernel-devel*noarch.rpm
+	kernelDefaultDevelPattern := fmt.Sprintf("kernel-default-devel-%s%s.rpm", kr.Fullversion, kr.FullExtraversion)
+	kernelDevelNoArchPattern := strings.ReplaceAll( // need to replace architecture string with "noarch"
+		fmt.Sprintf("kernel-devel-%s%s.rpm", kr.Fullversion, kr.FullExtraversion),
+		kr.Architecture.ToNonDeb(),
+		"noarch",
+	)
+
+	// get all possible URLs
+	possibleURLs := buildURLs(kr, kernelDefaultDevelPattern, kernelDevelNoArchPattern)
+
+	// trim the list to only resolving URLs
+	urls, err := getResolvingURLs(possibleURLs)
```

Comment:
Not necessarily for this PR, but do you think it's worthwhile to add a caching layer on top of `getResolvingUrls()`? For instance create a map `kernelrelease.KernelRelease[]string map` that would map a kernelrelease to the results of (expensive) calls to `getResolvingUrls()`. Users who run kernel-crawler/driverkit internally would benefit from not repeatedly issuing requests for URLs that will likely repeatedly fail.

Reply:
I had this same thought as well, and right now there's not really a good solution that I can think of... The problem with SUSE, at least as far as I can tell, is that they only host a single kernelrelease for a given distro release at a time. So you can't know beforehand where that kernel release is hosted on the myriad of sites (unless you have kernel crawler tell you, but that's still expensive to run). But when someone passes in `--kernelrelease 6.0.5-3.1.g7359656.x86_64`, there's no way to know (without a bunch of hardcoding) which openSUSE release that's for..

I think the ultimate solution is there needs to be a better way to discover the kernel package urls in general... Part of me wonders if the container-based approach and some way to query directly from the OS package manager might be better. 

But I knew this URL-based solution was faster to write and gets us a starting place. Another solution is to just remove the `getResolvingUrls()` entirely. The whole slew of possible URLs will be passed to the next bits of code, but 🤷 .